Userland/Libraries/LibGUI/RegularEditingEngine.cpp

```diff
@@ -57,10 +57,7 @@ void RegularEditingEngine::sort_selected_lines()
 
     auto& lines = m_editor->document().lines();
 
-    auto start = lines.begin() + (int)first_line;
-    auto end = lines.begin() + (int)last_line + 1;
-
-    quick_sort(start, end, [](auto& a, auto& b) {
+    dual_pivot_quick_sort(lines, static_cast<int>(first_line), static_cast<int>(last_line + 1), [](auto& a, auto& b) {
```

Comment:
While agree with the other changes in this commit,  `AK: Remove Iterator quick_sort ` to use the collections API, I'm not sure I understand why we would remove the iterator altogether? 

It clearly makes sense to call it for this case in LibGUI, where we don't want to sort the entire container, just the selected lines. I could see other use cases in e.g. `Mail` where we would want to sort part of a container. 

So, I would rather we drop this file and AK/QuickSort.h from this commit and just have it be a "use container quick sort in more places" PR.

Reply:
Sorting part of a container can always be achieved with start and end indices, which is equivalent (and most often more convenient) than sorting with iterators. Just because somewhere an API *could* be useful doesn't mean that we have to keep it, that's not how we do things I believe.

Reply:
The API that is explicitly exposed from `quick_sort` is `AK::quick_sort`, not `single_pivot_quick_sort` or `dual_pivot_quick_sort`.  

Whether the distinction that calling the iterator version calls single_pivot_quick_sort while the collection version calls dual_pivot_quick_sort makes any sense is out of scope here. (though personally, I don't think it makes any sense).

There are zero callers of single_pivot_quick_sort outside of Tests. 
There are two callers of dual_pivot_quick_sort outside of Tests. One is Assistant, the other is LibC `qsort`. 

APIs using collections are nicer than APIs using Iterators, and APIs using explict iterators are nicer than those using indices. Why would we remove a nicer API to tell callers to use a more internal one?

So, by changing from `quick_sort(Iterator, Iterator)` to `dual_pivot_quick_sort(collection, int, int)` you are both changing the behavior (from single to dual pivot) and using a worse API for callers.

Index-based and Iterator-based APIs are not equivalent. Remove the removal or find another maintainer to merge the changes.

Reply:
I see, it has been too long since I actually looked at the code closely. I'll just drop that commit then